### PR TITLE
[Messenger] Allows to register handlers on a specific transport

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30110
| License       | MIT
| Doc PR        | symfony/symfony-docs#11236

With the #30008 pull-request in, we can now do the following:
```yaml
framework:
    messenger:
        transports:
            events:
                dsn: amqp://guest:guest@127.0.0.1/%2f/events
                options:
                    queues:
                        3rdparty:
                            binding_keys: [ 3rdparty ]
                        projection:
                            binding_keys: [ projection ]

            events_3rdparty: amqp://guest:guest@127.0.0.1/%2f/events?queues[3rdparty]
            events_projection: amqp://guest:guest@127.0.0.1/%2f/events?queues[projection]

        routing:
            'App\Message\RegisterBet': events
``` 

This will bind two queues to the `events` exchange, fantastic:
Screenshot 2019-04-07 at 10 26 27

---

Now, in this setup, the message will be duplicated within the `3rdparty` & `projection` queues. If you just run the consumer for each transport, it will consume the message and call all the handlers. You can't do different things based on which queue you have consumed the message. **This pull-request adds the following feature:**

```php
class RegisterBetHandler implements MessageSubscriberInterface
{
    public function __invoke(RegisterBet $message)
    {
        // Do something only when the message comes from the `events_projection` transport...
    }

    /**
     * {@inheritdoc}
     */
    public static function getHandledMessages(): iterable
    {
        yield RegisterBet::class => [
            'from_transport' => 'events_projection',
        ];
    }
}
```

---

In the debugger, it looks like this:
Screenshot 2019-04-07 at 10 29 55

### CODE REVIEW

**Comment by Koc** [comment]
I see some drawbacks with current configuration:
* All this `queues[3rdparty][binding_keys][]=3rdparty&queues[projection][binding_keys][]=projection` dsn strings is quite hard to understand and write. And dsn string became longer and longer on each new queue.
* You should create +1 transport for consuming messages from each queue (`events_3rdparty`, `events_projection`)
* Handler (code level) knows about transport name (config level) - it similar to breaking dependency inversion principe but for config

Please consider move out queue name from dsn and extend routing configuration:

```yaml
framework:
    messenger:
        transports:
            events: amqp://guest:guest@127.0.0.1/%2f/events
        routing:
            App\Message\RegisterBet:
                -   transport: events
                    queue: events_projection
                    handler: App\MessageHandler\RegisterBetHandler
                -   transport: events
                    queue: events_3rdparty
                    handler: App\MessageHandler\SomethingOnBetRegister

```
Handler can be omit for cases when we just pulishing messages and don't comsuming they through Messenger.

This kind of config is super easy to understand. We configuration of message-transport/queue-handler in single place.

Of counse `messenger:consume` command should be updated to use handler name as argument `messenger:consume 'App\MessageHandler\RegisterBetHandler'`

**Comment by sroze** [comment]
@Koc I don't see your point as being against this proposed implementation here. The main point is to create this `HandlerDescriptor`, which allows us to have additional metadata on a specific handler (following a very long discussion with @webmozart). 

In the example given, the handler indeed knows about the transport (it's already knowing about _things_ like the `bus` and its own `priority` when using the `MessageSubscriberInterface`). But it allows us to harvest these options (`HandlerDescriptor`'s `options`) from outside now, which can indeed be a YAML configuration.

Regarding the overhead in terms of configuring the exchange and the queue bindings, you could also use the `options` if you don't want it in the DSN and might challenge would here be that actually you'd be better in creating this setup by yourself in RabbitMq. Having to create multiple transports actually simplifies the configuration because we don't need to introduce this extra concept `queue` that you suggested above.

**Comment by sroze** [comment]
(requires CHANGELOG when a few 👍 will be around 😉 )

**Comment by sroze** [comment]
Open question: should we only have `handlerName` in `HandlerDescriptor` instead of directly the handler so it can later be serialized if needed?

**Comment by weaverryan** [comment]
Makes sense to me - we do this with the bus name stamp and sender name for retry.

File: src/Symfony/Component/Messenger/Command/DebugCommand.php [hunk 1]
```diff
@@ -97,4 +99,18 @@ protected function execute(InputInterface $input, OutputInterface $output)
             }
         }
     }
+
+    private function formatConditions(array $options): string
+    {
+        if (empty($options)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Command/DebugCommand.php]: `if (!$options) {`

File: src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ */
+
+namespace Symfony\Component\Messenger\Handler;
+
+/**
+ * Describe a handler and the possible associated options, such as `transport`, `bus`, etc...
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ *
+ * @final
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php]: let's make it really `final`

File: src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Handler;
+
+/**
+ * Describe a handler and the possible associated options, such as `transport`, `bus`, etc...
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php]: `etc.` (ellipses are redundant)

**Review by nicolas-grekas — COMMENTED** [review]
LGTM, just a few minor things + rebase needed

**Comment by sroze** [comment]
@nicolas-grekas @weaverryan updated 👍

File: src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Handler;
+
+/**
+ * Describe a handler and the possible associated options, such as `transport`, `bus`, etc.
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php]: `@experimental in 4.3`

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 5]
```diff
@@ -80,6 +80,11 @@ CHANGELOG
  * Added a Doctrine transport. For example, use the `doctrine://default` DSN (this uses the `default` Doctrine entity manager)
  * [BC BREAK] The `getConnectionConfiguration` method on Amqp's `Connection` has been removed. 
  * [BC BREAK] A `HandlerFailedException` exception will be thrown if one or more handler fails.
+ * [BC BREAK] The `HandlersLocationInterface::getHandlers` method needs to return `HandlerDescriptor`
+   instances instead of callables.
+ * [BC BREAK] The `handlerAlias` has been renamed to `handlerName` on the `HandledStamp`.
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/CHANGELOG.md]: `getCallableName` is gone and the constructor changed as well.

**Review by weaverryan — COMMENTED** [review]
This looks good to me - and I tested it locally.

There's just one piece that bothers me:

```php
    public static function getHandledMessages(): iterable
    {
        yield RegisterBet::class => [
            'transport' => 'events_projection',
        ];
    }
```

The *meaning* of `transport` is this:

> This handler will only be called when the message is received from the `events_projection` transport

But it's just as reasonable to think it means this:

> This handler will be sent to the `events_projection` transport

But, the only possibly better key I can think of is: `'only_handle_from_transport' => 'events_projection',`

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 6]
```diff
@@ -94,32 +95,33 @@ private function registerHandlers(ContainerBuilder $container, array $busIds)
...
-                                $messageLocation = isset($tag['handles']) ? 'declared in your tag attribute "handles"' : $r->implementsInterface(MessageSubscriberInterface::class) ? sprintf('returned by method "%s::getHandledMessages()"', $r->getName()) : sprintf('used as argument type in method "%s::%s()"', $r->getName(), $method);
+                    $priority = $tag['priority'] ?? $options['priority'] ?? 0;
+                    $method = $options['method'] ?? '__invoke';
 
-                                throw new RuntimeException(sprintf('Invalid configuration %s for message "%s": bus "%s" does not exist.', $messageLocation, $message, $method['bus']));
-                            }
+                    if (isset($options['bus'])) {
+                        if (!\in_array($options['bus'], $busIds)) {
+                            $messageLocation = isset($tag['handles']) ? 'declared in your tag attribute "handles"' : $r->implementsInterface(MessageSubscriberInterface::class) ? sprintf('returned by method "%s::getHandledMessages()"', $r->getName()) : sprintf('used as argument type in method "%s::%s()"', $r->getName(), $method);
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: don't forget to take https://github.com/symfony/symfony/pull/31252 into account :)

**Comment by chalasr** [comment]
> But, the only possibly better key I can think of is: 'only_handle_from_transport' => 'events_projection',

Maybe `from_transport` is enough to avoid the confusion?

**Comment by weaverryan** [comment]
Hmm, I do kinda like `from_transport`

**Comment by sroze** [comment]
Happy with `from_transport` as well. Let me change this 👍

**Comment by sroze** [comment]
This PR should be ready :)

File: src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Handler;
+
+/**
+ * Describe a handler and the possible associated options, such as `from_transport`, `bus`, etc.
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php]: Describes

**Review by chalasr — APPROVED** [review]
:+1: with minor comment

File: src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php (hunk 7 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Handler/HandlerDescriptor.php]: Nice catch 👍

**Comment by sroze** [comment]
(failure is unrelated)

**Review by nicolas-grekas — APPROVED** [review]

**Comment by Tobion** [comment]
The `from_transport` is not documented in the `MessageSubscriberInterface`.

I still can't believe we use arbitrary array values in the suscriber instead of a value object that is self-documenting and can have types. There is even a `HandlerDescriptor` now but only for internal use...

**Comment by Bavragor** [comment]
Please add the wonderful example from @sroze to the documentation on https://symfony.com/doc/current/messenger.html as a how to, took me some time to find this pull request. :)

**Pull request merged**